### PR TITLE
Add date input `day`, `month`, `year` and `values` options

### LINKED ISSUE #1154: Pass values to date-input when using default items
At the moment there is no way to pass values or field-level error states (CSS classes) to the date-input when using the default item set. This means you have to redefine the items param, including CSS classes for widths, whenever it's used. This encourages repeated code and is arguably more prone to error.

Is it worth having a values param and an errors param on date-input? This could fall back to values/errors passed via the `items` param for backward compatibility, but favouring the new `items`/`errors` props.

e.g.
```nunjucks
{{ govukDateInput({
  id: "dob",
  namePrefix: "dob",
  values: {
    day: 4,
    month: 5,
    year: 20t19
  },
  errorMessage: 'Please enter a valid year',
  errors: {
    year: true
  }
}) }}
```

I don't have an immediate need for this but it's something I noticed and thought worth suggesting in case others agree.

### PR DESCRIPTION
**Pull request by colinrotherham**
This PR contributes some date input changes from NHS.UK frontend to add new Nunjucks options:

- `day`, `month` and `year` to customise individual items
- `error` boolean option to set the error state on individual items
- `values` option to set individual item values using a single object

It closes https://github.com/alphagov/govuk-frontend/issues/1154 and is based on: https://github.com/nhsuk/nhsuk-frontend/pull/1869

Apologies to @joelanman as this PR re-implements https://github.com/alphagov/govuk-frontend/pull/4932 without any breaking changes

## Setting individual values and errors

The date input error state is now set automatically when `errorMessage` is provided.

For example, it is no longer necessary to set the `"govuk-input--error"` class on every item:

```patch
  {{ govukDateInput({
    fieldset: {
      legend: {
        text: "What is your date of birth?"
      }
    },
    errorMessage: {
      text: "Enter your date of birth"
-   },
-   items: [
-     {
-       name: "day",
-       label: "Day",
-       classes: "govuk-input--width-2 govuk-input--error"
-     },
-     {
-       name: "month",
-       label: "Month",
-       classes: "govuk-input--width-2 govuk-input--error"
-     },
-     {
-       name: "year",
-       label: "Year",
-       classes: "govuk-input--width-4 govuk-input--error"
-     }
-   ]
+   }
  }) }}
```

If only one field has an error, use the new `day`, `month` or `year` options to set `error: true` for that field only:

```patch
  {{ govukDateInput({
    fieldset: {
      legend: {
        text: "What is your date of birth?"
      }
    },
    errorMessage: {
      text: "Date of birth must include a year"
    },
-   items: [
-     {
-       name: "day",
-       label: "Day",
-       classes: "govuk-input--width-2"
-     },
-     {
-       name: "month",
-       label: "Month",
-       classes: "govuk-input--width-2"
-     },
-     {
-       name: "year",
-       label: "Year",
-       classes: "govuk-input--width-4 govuk-input--error"
-     }
-   ]
+   year: {
+     error: true
+   }
  }) }}
```

Similarly, you can also use `error: false` to _remove_ a single red border versus multiple `error: true`

For example, it can make it easier to turn off the error state for the fields you have values for:

```patch
  {{ govukDateInput({
    fieldset: {
      legend: {
        text: "What is your date of birth?"
      }
    },
    errorMessage: {
      text: "Date of birth must include a month and year"
    },
    day: {
-     value: "31"
+     value: "31",
+     error: false
-   },
-   month: {
-     error: true
-   },
-   year: {
-     error: true
    }
  }) }}
```

## Setting all values at once

Or when using the GOV.UK Prototype Kit, given the following `data` object:

```json
{
  "dob-day": "31",
  "dob-month": "3",
  "dob-year": "1980"
}
```

You can now automatically set item values by passing in `values: data` as shown:

```njk
  {{ govukDateInput({
    fieldset: {
      legend: {
        text: "What is your date of birth?"
      }
    },
    namePrefix: "dob",
    values: data
  }) }}
```

### CODE REVIEW

**Comment by owenatgov** [comment]
Hi @colinrotherham. This is looking great! Thanks for reimplementing this.

At the moment we're keen to keep `main` ring-fenced whilst we move Frontend to sass modules, for which we're planning on doing a beta release within the next couple of weeks which we want out in the world for a little while to test for any issues with our implementation. After that, pending review, we should be able to bring this in.

I'll run this past the other devs but if they're happy with this solution we can reivew this and put it on hold until we're ready to merge it.

**Comment by colinrotherham** [comment]
@owenatgov Need me to rebase?

**Comment by owenatgov** [comment]
@colinrotherham I'm just reviewing now but I was gonna mention yep we'll want a rebase to fix that changelog conflict. If you're up for it then go right ahead.

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml [hunk 1]
```diff
@@ -28,6 +28,10 @@ params:
         type: string
         required: false
         description: If provided, it will be used as the initial value of the input.
+      - name: error
+        type: boolean
+        required: false
+        description: If set to `true`, show a red border on the item input.'
```
Comment (owenatgov) [inline packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: I like the format of this as part of a more nuanced error system for the date input but I think this description needs to be clearer about what it does and what it's for. We may want to bring a tech writer in but my rough suggestion would be:


```suggestion
        description: If set to `true`, will show a red error border on the item input. Should be used in tandem with the `errorMessage` attribute to provide a clear error message for the whole component and a visual error indicator for the specifically effected item(s).'
```

This also gets rid of a stray apostrophe at the end of the description

**Review by owenatgov — COMMENTED** [review]
This is looking very nice. I've left a comment about the wording of the `error` option to make it clearer but other than that I'm pretty happy with code.

You know the drill that this then needs a second developer review. I also think it's worth bringing a tech writer in to give the changelog and new options the once over.

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 1 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: Checkboxes, radios and select talk about a top-level `value` or `values` option.

Shall we use the same wording here for top-level `errorMessage` option`?

Comment (joelanman) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: My suggestion
```
If `true`, shows a red border on the item input. You must also use the `errorMessage` option to display an error message.
```
is `errorMessage` on the same object as `error`?

edit - oh ha posted at the same time as @colinrotherham

Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: How about something a bit shorter?

Mainly because `classes: "govuk-input--error"` currently lacks any strict guidance

```suggestion
        description: If set to `true`, will show a red error border on the item input. If provided, you must set the top-level `errorMessage` option to add an error message to the date input component.
```

**Comment by colinrotherham** [comment]
>@colinrotherham I'm just reviewing now but I was gonna mention yep we'll want a rebase to fix that changelog conflict. If you're up for it then go right ahead.

Rebased and pushed, thanks

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 1 above)
Comment (owenatgov) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: I'm happy with the less verbose version 👍🏻

Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: Brill, pushed that one up

**Review by owenatgov — APPROVED** [review]
Great!

Gonna prompt the other DS developers for a second review, plus our tech writer on Monday

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml [hunk 2]
```diff
@@ -132,6 +201,16 @@ examples:
...
   - name: month and year
+    options:
+      id: dob
+      namePrefix: dob
+      fieldset:
+        legend:
+          text: When did you move to this property?
+      hint:
+        text: For example, 3 1980
+      year: false
```
Comment (36degrees) [inline packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: Based on the example name it should be `day` that's omitted, not `year`.

```suggestion
- name: month and year
    options:
      id: dob
      namePrefix: dob
      fieldset:
        legend:
          text: When did you move to this property?
      hint:
        text: For example, 3 1980
      day: false
```

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk [hunk 3]
```diff
@@ -11,25 +11,92 @@
...
+    dateInputYear
   ] %}
 {% endif %}
 
+{%- macro _dateInputItem(params, item, anyItemHasError) %}
+{% set itemName = item.name %}
+{% set itemValue = item.value %}
+{% set itemWidth = 2 %}
+{% set itemClasses = "" %}
+{% set itemHasError = true if item.error or (item.classes and "govuk-input--error" in item.classes) else false %}
```
Comment (36degrees) [inline packages/govuk-frontend/src/govuk/components/date-input/template.njk:47]: We don't need to take the existence of the `"govuk-input--error"` class in `item.classes` into account here, because the only thing `itemHasError` controls is whether `govuk-input--error` gets added to `itemClasses`, and that'll happen anyway when we append `item.classes` to `itemClasses`.

(As currently implemented, setting e.g. `classes: govuk-input--error` on `year` results in the class being included _twice_)

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk [hunk 4]
```diff
@@ -11,25 +11,92 @@
...
-      name: "year",
-      classes: "govuk-input--width-4"
-    }
+    dateInputDay,
+    dateInputMonth,
+    dateInputYear
   ] %}
 {% endif %}
 
+{%- macro _dateInputItem(params, item, anyItemHasError) %}
```
Comment (36degrees) [inline packages/govuk-frontend/src/govuk/components/date-input/template.njk:42]: The `anyItemHasError` parameter and the logic for setting the `anyItemHasError` on line 94 is introduced in 35e7001e130c0f294e3fe955461c4cbdd0587e7d but isn't used until 3f31571bd8b21c9a9c78525c7e22a7874d3d718b – can we sort the history out so that it makes sense?

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk [hunk 5]
```diff
@@ -11,25 +11,92 @@
...
+    inputmode: item.inputmode if item.inputmode else "numeric",
+    autocomplete: item.autocomplete,
+    pattern: item.pattern,
+    attributes: item.attributes
+  }) | trim | indent(2) }}
+</div>
+{%- endmacro -%}
+
+{#- Determine if any date input item has an error #}
+{%- set anyItemHasError = false %}
```
Comment (36degrees) [inline packages/govuk-frontend/src/govuk/components/date-input/template.njk:97]: It doesn't look like we have any coverage for applying the input error state when no items have error or the `govuk-input--error` class. Can we make sure we have tests to cover:

- all inputs should be in error state if none of the inputs have `error: true` or a `govuk-input--error` class
- other inputs should not be in an error state if `error` is `true` on one input
- other inputs should not be in an error state if `classes` contains `govuk-input--error` on one input
- inputs should not be in error state if `error: false` and none of the inputs have `error: true` or a `govuk-input--error` class

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,112 @@ For advice on how to use these release notes, see [our guidance on staying up to
 
 ## Unreleased
 
+### New features
+
+#### Add date input `day`, `month`, `year` and `values` options
```
Comment (36degrees) [inline CHANGELOG.md]: I'd suggest waiting until @seaemsi has had a chance to review this too, but I think this should be user-focused rather than describe the change that we made.

So e.g. 'You can now pass `day`, `month`, `year`, `error` and `values` options to the date input component'

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk (hunk 5 above)
Comment (36degrees) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:97]: I think this might be another place where the history is a little confusing, I think there are tests for this but they were included in 35e7001, before `anyItemHasError` is implemented in 3f31571.

I still think it'd be beneficial to have separate tests with a clear description, the 'can have classes for individual items' test is testing too many things and isn't explicit about what it's testing.

**Review by 36degrees — COMMENTED** [review]
Thanks @colinrotherham, this looks like a great improvement for users 🙌🏻 

I have flagged a couple of places where I don't think we have enough test coverage. The code is getting pretty complicated, so I think having comprehensive and descriptive tests for all the various edge-cases the code is trying to handle is going to be key to maintaining this.

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,112 @@ For advice on how to use these release notes, see [our guidance on staying up to
...
+
+We've updated the date input component to add new Nunjucks options:
```
Comment (seaemsi) [inline CHANGELOG.md]:
```suggestion
We've updated the Date input component to add the following Nunjucks options:
```

File: CHANGELOG.md (hunk 6 above)
Comment (seaemsi) [inline CHANGELOG.md]:
```suggestion
#### You can now pass `day`, `month`, `year`, `error` and `values` options to the Date input component
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,112 @@ For advice on how to use these release notes, see [our guidance on staying up to
...
+
+- `day`, `month` and `year` to customise individual items
+- `error` boolean option to set the error state on individual items
+- `values` option to set individual item values using a single object
+
+For consistency with other components, the date input error state is now set automatically when `errorMessage` is provided.
```
Comment (seaemsi) [inline CHANGELOG.md]:
```suggestion
For consistency with other components, the Date input component's error state is now set automatically when `errorMessage` is provided.
```

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml [hunk 9]
```diff
@@ -28,6 +28,10 @@ params:
         type: string
         required: false
         description: If provided, it will be used as the initial value of the input.
+      - name: error
+        type: boolean
+        required: false
+        description: If set to `true`, will show a red error border on the item input. If provided, you must set the top-level `errorMessage` option to add an error message to the date input component.
```
Comment (seaemsi) [inline packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]:
```suggestion
        description: If set to `true`, it will show a red error border on the item input. If provided, you must set the top-level `errorMessage` option to add an error message to the Date input component.
```

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml [hunk 10]
```diff
@@ -98,6 +102,35 @@ params:
...
+    required: false
+    description: Options for the month input within the date input component. See items.
+  - name: year
+    type: object
+    required: false
+    description: Options for the year input within the date input component. See items.
+  - name: values
+    type: object
+    required: false
+    description: An optional object use to specify `value` attributes for the date parts without setting items.
```
Comment (seaemsi) [inline packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]:
```suggestion
    description: An optional object used to specify `value` attributes for the date parts without setting items.
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,112 @@ For advice on how to use these release notes, see [our guidance on staying up to
 
 ## Unreleased
 
+### New features
+
+#### Add date input `day`, `month`, `year` and `values` options
+
+We've updated the date input component to add new Nunjucks options:
+
+- `day`, `month` and `year` to customise individual items
```
Comment (seaemsi) [inline CHANGELOG.md]:
```suggestion
- `day`, `month` and `year` options to customise individual items
```

**Review by seaemsi — COMMENTED** [review]
Looks good! Have made some style/content suggestions.

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk [hunk 12]
```diff
@@ -11,25 +11,92 @@
...
+<div class="govuk-date-input__item">
+  {{ govukInput({
+    label: {
+      text: item.label if item.label else itemName | capitalize,
+      classes: "govuk-date-input__label"
+    },
+    id: item.id if item.id else (params.id + "-" + itemName),
+    classes: "govuk-date-input__input" + (" " + itemClasses if itemClasses),
+    name: namePrefix + itemName,
+    value: itemValue | default(params.values[namePrefix + itemName]),
```
Comment (36degrees) [inline packages/govuk-frontend/src/govuk/components/date-input/template.njk:86]: There's no test coverage for the `default(params.values[namePrefix + itemName])` behaviour – can we add an example that uses `namePrefix` and prefixed `values` and a test for the behaviour?

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 10 above)
Comment (36degrees) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: I think we should also document that the keys in the `values` object can also be prefixed to match `namePrefix`, somehow.

**Review by 36degrees — COMMENTED** [review]
I've also realised that we have no tests or docs for name-prefixed values (other than the Prototype Kit focused guidance in the release notes).

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 2 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: Brilliant, thanks. All sorted

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk (hunk 3 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:47]: The variable `itemHasError` should always be correct, error boolean or error class

I've added a check for `not item.classes or not "govuk-input--error" in item.classes` though before it gets added to `itemClasses` (thanks), so it now avoids the duplicate error class

Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:47]: It's omitted it from this PR but your users might find `width` useful too:

```patch
  year: {
-   classes: 'govuk-input--width-2'
+   width: 2
  }
```

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk (hunk 4 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:42]: Sorted, thanks

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk (hunk 5 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:97]: Coverage added for all of these

File: CHANGELOG.md (hunk 6 above)
Comment (colinrotherham) [reply CHANGELOG.md]: Done 👍

File: CHANGELOG.md (hunk 7 above)
Comment (colinrotherham) [reply CHANGELOG.md]: Done 👍

File: CHANGELOG.md (hunk 6 above)
Comment (colinrotherham) [reply CHANGELOG.md]: Done 👍

File: CHANGELOG.md (hunk 8 above)
Comment (colinrotherham) [reply CHANGELOG.md]: Changed this, thanks

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 9 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: I've changed this but ignored the capitalisation change

There are 11x all-lowercase "date input component" in this YAML file (and all the other components)

Probably leave that to you to fix them all in another PR, rather than just this one of the 11

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk (hunk 4 above)
Comment (36degrees) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:42]: Thanks! 🙏🏻

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 10 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: Fixed use → used

>I think we should also document that the keys in the `values` object can also be prefixed to match `namePrefix`, somehow.

☝️ Will leave this one to you and the team

File: CHANGELOG.md (hunk 11 above)
Comment (colinrotherham) [reply CHANGELOG.md]: Done, thanks

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk (hunk 12 above)
Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:86]: Added this

**Comment by colinrotherham** [comment]
There was no conflict yesterday, rebasing again 😆

**Comment by owenatgov** [comment]
@colinrotherham Cheers for addressing everything! Unfortunately by chance we're in an unofficial freeze at the moment in prep for releasing 6.3.0 this morning. Once that's out we'll give this a re-review. All being well this'll go out in 6.4.0. I'll pin it to that milestone now.

File: packages/govuk-frontend/src/govuk/components/date-input/template.njk [hunk 13]
```diff
@@ -11,25 +11,94 @@
...
+{% endif %}
+
+{%- set namePrefix = (params.namePrefix + "-") if params.namePrefix else "" -%}
+
+<div class="govuk-date-input__item">
+  {{ govukInput({
+    label: {
+      text: item.label if item.label else itemName | capitalize,
+      classes: "govuk-date-input__label"
+    },
```
Comment (colinrotherham) [inline packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: @36degrees Sorry, just remembered something else

Unlike other components you can't use `label.text` or `label.html` in date input fields

Only a plain text `label` is supported

This can be fixed in a backwards compatible way though if you want it adding?

```njk
{% set itemLabel = item.label if item.label is mapping and item.label is not escaped else {
  text: item.label if item.label is string else undefined,
  html: item.label if item.label is escaped else undefined
} %}
```

Comment (36degrees) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: This feels like a separate piece of work, so let's leave it out of this PR. I think we'd also need to work out how to document this as label could then either be a string or an object.

Are you aware of any situations where this might actually be a problem? Or is this more about consistency? I don't recall it ever coming up as an issue.

Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: Nah it's not a problem because you can always do `label: customHtml | safe`

Similar to the error classes not being automatic, it just stood out as different to other components

E.g. If you needed to preserve HTML entities in localised labels for whatever reason

Comment (joelanman) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: just to note I'm not sure this logic is correct? If only text is supported, you can't get html through, it will be escaped

Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: Which bit @joelanman, the workaround to pass HTML into `params.text` using `| safe`?

Comment (joelanman) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: yeh, though maybe I dont understand how safe works!

Comment (colinrotherham) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: Preferably the Nunjucks API would have predictable `text` and `html` options, but yeah `| safe` can be used if you absolutely have to. It skips auto escaping.

Regarding `is not escaped` above, that's because in Nunjucks a `| safe` escaped string becomes an [instance of `SafeString`](https://github.com/mozilla/nunjucks/blob/v3.2.4/nunjucks/src/runtime.js#L156-L159) and (as an object) incorrectly [passes the `is mapping` test](https://github.com/mozilla/nunjucks/blob/v3.2.4/nunjucks/src/tests.js#L271-L288)

But used together, it's how `govukAttributes()` detects strings versus objects

Great for backwards compatibility

Comment (36degrees) [reply packages/govuk-frontend/src/govuk/components/date-input/template.njk:82]: I've created #7202 to capture this

**Review by seaemsi — APPROVED** [review]

File: packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml (hunk 10 above)
Comment (36degrees) [reply packages/govuk-frontend/src/govuk/components/date-input/date-input.yaml]: Added in 7be6322.

**Comment by 36degrees** [comment]
@seaemsi would you mind one final review of just the last commit I pushed, which resolves [this issue](https://github.com/alphagov/govuk-frontend/pull/6971#discussion_r3386769898).

**Review by 36degrees — APPROVED** [review]

**Review by seaemsi — APPROVED** [review]
